src/main/java/io/github/thebusybiscuit/slimefun4/core/services/MetricsService.java

```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * Gets the latest version available as an int.
+     * This is an internal method used by {@link #checkForUpdate(String)}.
+     * If it cannot get the version for whatever reason this will return 0, effectively always
+     * being behind.
+     *
+     * @return The latest version as an integer or -1 if it failed to fetch.
+     */
+    private int getLatestVersion() {
+        try {
+            HttpResponse<JsonNode> response = Unirest.get(GH_API + "/releases/latest")
+                .asJson();
+            if (!response.isSuccess()) return -1;
+
+            JsonNode node = response.getBody();
+
+            if (node == null) return -1;
+
+            return node.getObject().getInt("tag_name");
+        } catch (UnirestException e) {
+            plugin.getLogger().log(Level.SEVERE, "Failed to fetch latest builds for SFMetrics");
+            return -1;
+        }
+    }
+
+    /**
+     * Downloads the version specified to Slimefun's data folder.
+     *
+     * @param version The version to download.
+     */
+    private boolean download(int version) {
+        File f = new File(plugin.getDataFolder(), "Metrics-" + version + ".jar");
+
+        try {
+            info("# Starting download of MetricsModule build: #" + version);
+            AtomicInteger lastPercentPosted = new AtomicInteger();
+            HttpResponse<File> response = Unirest.get(GH_REPO_RELEASES + "/" + version
+                + "/" + REPO_NAME + ".jar")
+                .downloadMonitor((b, fileName, bytesWritten, totalBytes) -> {
+                    int percent = (int) (20 * (Math.round((((double) bytesWritten / totalBytes) * 100) / 20)));
+
+                    if (percent != 0 && percent != lastPercentPosted.get()) {
+                        info("# Downloading... " + percent + "% " +
+                            "(" + bytesWritten + "/" + totalBytes + " bytes)");
+                        lastPercentPosted.set(percent);
+                    }
+                })
+                .asFile(f.getPath());
+            if (response.isSuccess()) {
+                info("Successfully downloaded " + REPO_NAME + " build: " + version);
+
+                // Replace the metric file with the new one
+                Files.move(f.toPath(), metricFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+
+                metricVersion = String.valueOf(version);
+                return true;
+            }
+        } catch (UnirestException e) {
+            plugin.getLogger().log(Level.WARNING, "Failed to fetch the latest jar file from the" +
+                " builds page. Perhaps GitHub is down.");
+        } catch (IOException e) {
+            plugin.getLogger().log(Level.WARNING, "Failed to replace the old metric file with the " +
+                "new one. Please do this manually! Error: {0}", e.getMessage());
+        }
+        return false;
+    }
+
+    private void info(String str) {
+        plugin.getLogger().info(str);
+    }
```

Comment:
Is this method really really necessary here?

Reply:
Necessary? No but useful? yes. This code has changed a lot, it saved a lot of code.